tools/gfx/d3d12/render-d3d12.cpp

```diff
@@ -752,3283 +580,4160 @@ class D3D12Device : public GraphicsAPIRenderer
+                viewDesc.SizeInBytes =
+                    (UINT)D3DUtil::calcAligned((UInt)m_ordinaryData.getCount(), 256);
+                encoder->m_device->CreateConstantBufferView(
+                    &viewDesc,
+                    m_resourceHeap->getCpuHandle(descriptorTable));
+            }
 
-                auto& cpuHeap = *descriptorSet->m_samplerHeap;
-                auto cpuDescriptorTable = descriptorSet->m_samplerTable;
+            return SLANG_OK;
+        }
 
-                m_device->CopyDescriptorsSimple(
-                    UINT(descriptorCount),
-                    gpuHeap.getCpuHandle(gpuDescriptorTable),
-                    cpuHeap.getCpuHandle(int(cpuDescriptorTable)),
+    public:
+        virtual Result bindObject(PipelineCommandEncoder* encoder, RootBindingState* bindingState)
+        {
+            ShaderObjectLayoutImpl* layout = getLayout();
+            SLANG_RETURN_ON_FAIL(_bindOrdinaryDataBufferIfNeeded(encoder));
+            uint32_t descTableIndex = bindingState->rootParamIndex;
+            auto& descSet = m_descriptorSet;
+            if (descSet.m_resourceCount)
+            {
+                auto gpuDescriptorTable = bindingState->descriptorTables[descTableIndex];
+                auto& gpuHeap = gpuDescriptorTable.heap;
+                auto& cpuHeap = *m_resourceHeap;
+                auto cpuDescriptorTable = descSet.m_resourceTable;
+
+                bindingState->device->m_device->CopyDescriptorsSimple(
+                    UINT(descSet.m_resourceCount),
+                    gpuHeap.getCpuHandle(gpuDescriptorTable.table + bindingState->offset.resource),
+                    cpuHeap.getCpuHandle(cpuDescriptorTable),
+                    D3D12_DESCRIPTOR_HEAP_TYPE_CBV_SRV_UAV);
+                bindingState->offset.resource += descSet.m_resourceCount;
+                descTableIndex++;
+            }
+            if (descSet.m_samplerCount)
+            {
+                auto gpuDescriptorTable = bindingState->descriptorTables[descTableIndex];
+                auto& gpuHeap = gpuDescriptorTable.heap;
+                auto& cpuHeap = *m_samplerHeap;
+                auto cpuDescriptorTable = (int)descSet.m_samplerTable;
+
+                bindingState->device->m_device->CopyDescriptorsSimple(
+                    UINT(descSet.m_samplerCount),
+                    gpuHeap.getCpuHandle(gpuDescriptorTable.table + bindingState->offset.sampler),
+                    cpuHeap.getCpuHandle(cpuDescriptorTable),
                     D3D12_DESCRIPTOR_HEAP_TYPE_SAMPLER);
-
-                submitter->setRootDescriptorTable(
-                    int(rootParameterIndex++), gpuHeap.getGpuHandle(gpuDescriptorTable));
+                bindingState->offset.sampler += descSet.m_samplerCount;
+                descTableIndex++;
+            }
+            bindingState->futureRootParamOffset =
+                Math::Max(descTableIndex, bindingState->futureRootParamOffset);
```

Comment:
I think I see what is going on here, and in my own versions of this code (that is, what I've written for Falcor in the past) I think I handled the same issue by just having two different recursive walks. Well, three actually:

* Case (1) is the recursive walk is to write the uniform/ordinary data of a shader object. This one copies the directly-contained data and then recurses on any interface-type sub-objects.

* Case (2) is the recursive walk is to write descriptors *into* descriptor a descriptor set (or in the case of D3D12, a pair of descriptor sets). Case (2a) writes the default CBV binding (if needed), filling it in using (1), and then delegates to case (2b) which writes all the directly-contained SRV/UAV/sampler descriptors. This path recurses on interface-type sub-objects via (2b) and on `ConstantBuffer<X>`-type sub-objects via (2a).

* Case (3) is a recursive walk to write whole descriptor sets. It creates the descriptor set(s) for the current object and then relies on (2a) to fill it in. Then it recurses on any `ParameterBlock<X>`-type sub-objects.

With that decomposition, I think case (3) can just keep a running counter of how many root parameters have been filled in so far and it will do just fine.